feat: add build documentation

matthew-chirgwin opened the pull request:
- Offers build documentation, with code to follow in another PR
- Covers the approach, tools used, the configuration, and end output for UI code
- A follow up PR will add (once agreed) how the built UI code is then built/deployed into the wider Strimzi

matthew-chirgwin committed:
feat: initial draft of build documentation/config

- Add 1st version of build documentation

Contributes to: strimzi/strimzi-ui#1

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
refactor: fix md styling issues

- fixes a few .md syntax/styling issues

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

samuel-hawker commented:
This looks good to me! Well thought out and all of the choices make sense to me, though I wouldn't want to approve until some of the other working developers on the UI approve saying they like all the choices you have made here. :)

matthew-chirgwin commented:
> This looks good to me! Well thought out and all of the choices make sense to me, though I wouldn't want to approve until some of the other working developers on the UI approve saying they like all the choices you have made here. :)

Thanks @samuel-hawker - I have requested a review from them already :)

nictownsend commented on docs/Build.md:
```diff
@@ -0,0 +1,145 @@
+
+| Option        | Value           | Purpose  |
+| ------------- | ------------- | -----:|
+| entry      | `client/Bootstrap/index.js` | Build entry point. Note the `scss/css` is expected to be imported by individual view layer code, which is imported (directly or indirectly) from this file  |
```
Casual mention over scss/css imports - is that (to be) mentioned elsewhere?

matthew-chirgwin replied:
I was going to be explicit in a 'Code Style' doc PR, but my suggestion would be a (s)css import per component (as a part of a given view layer). The CSS optimisation plugin will then deduplicate down to a single set of class names. This should keep builds fast when developing (ie just one file changing) but also minimal when built in production. I will update the doc to mention details to follow - plus a few other small tweaks.

nictownsend replied:
Will optimisation namespace where duplicates happen? Or fail build (hopefully this)?

matthew-chirgwin replied:
OOTB, it will merge:
File one:

```
.foo {
color: red;
height: 100%;
}
```
File two:
```
.foo {
color: green;
}
.bar: {
color: red;
}
```

will become:

```
.foo {
color: green;
height: 100%;
}
.bar: {
color: red;
}
```

when built, with standard css precedence (ie later takes precedence) based on import order.

If it is ok, I will look at options here, and if we want to fail on conflicts - thoughts welcome. I can see pro's and cons either way. Based on discussion, I will update the documentation here with our decision in a follow up PR. Worth noting this optimisation only applies on production builds - when in development, we would see multiple (identical) class names (say), and where they originate from.

nictownsend replied:
Yep sounds good. My main concern is over whether we have to enforce name-spacing as a code style, or it can be explicitly forced.

matthew-chirgwin committed:
refactor: add extra details given reviews

- adds additional details given reviews
- tweaks config values

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
fix: spelling, add final details re config/aliases

- fixes a spelling error, and adds details around configuration
arrangement and defined code aliases

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

scholzj reviewed (commented):
I don't really understand much about the content. But ping me when you think it i sready - I can approve and merge it ;-)

seanforyou23 reviewed (approved):
Nice work @matthew-chirgwin! This looks great 😎

matthew-chirgwin commented:
Thank you for reviews @nictownsend and @seanforyou23 ! I think we are all good to merge now @scholzj - a PR to implement the above will follow shortly. @nictownsend / @seanforyou23 would you mind reviewing that when it is raised?

ppatierno reviewed (approved):
Everything seems to be fine as @matthew-chirgwin said :-) so I am happy to approve!

scholzj commented:
@samuel-hawker Do you have anything more? Or can I merge it? Thanks.

samuel-hawker commented:
Nothing more to add from me, this is a great start!
